File: 69-sqrt(x).cpp
int mySqrt(int x) {
    if (x == 0) return 0;
    int l = 1, r = x;
    int mid, sqrt;
    while (l <= r) {
        mid = l + (r - l) / 2;
        sqrt = x / mid;
        if (sqrt == mid) return mid;
        else if (sqrt < mid) r = mid - 1;
        else l = mid + 1;
    }
    return r;
}

File: python version/69-sqrt(x).py
class Solution:
    def mySqrt(self, x: int) -> int:
        if x == 0: return 0
        l, r = 1, x
        while l < r:
            # 思考最后只有两个数字的情况，需要强制向上取整
            mid = l + (r-l)//2 + 1
            if mid * mid == x: return mid
            elif mid * mid < x: l = mid
            else: r = mid - 1

        return l